test/utils.py

```diff
@@ -0,0 +1,373 @@
+
+    def update_output_status(self):
+        try:
+            result = self.results_out.get(False)
+        except queue.Empty:
+            return
+        self.waiting = True
+        key, *_ = result
+        self.pool.tasks[key].output = result
+
+    @classmethod
+    def _worker(cls, tasks_in: mp.Queue, results_out: mp.Queue):
+        while True:
+            try:
+                task = tasks_in.get()
+            except (EOFError, ValueError):
+                return
+            if task is None:
+                return  # Sentinel value, stop
+            key, fn, args, kwargs = task
+            try:
+                result = key, True, fn(*args, **kwargs)
+            except Exception as e:
+                result = key, False, mp.pool.ExceptionWithTraceback(e)
+            try:
+                results_out.put(result)
+            except Exception as e:
+                results_out.put(mp.pool.MaybeEncodingError(e, result[1]))
+
+
+async def _run_with_timeout_async_process(timeout: float, fn, args, kwargs,
+                                          debug=0, interval=0.005):
+    """fn must be pickleable and a regular function **not** coroutine!
+
+    debug:
+      - 0: normal
+      - 1: print pid on start
+      - 2: also wait 15s in process to allow time to 'Attach to Process' in Pycharm
+    Note: don't use debug=2 with low timeouts unless you want to get terminate()d"""
+    dest = mp.Queue()
+    # The daemon's main purpose is to stop ourselves accidentally creating a fork-bomb
+    #  by running run_with_timeout in the worker recursively
+    #  (we do weird stuff to get pickle to work which might break?).
+    # This is because daemonic processes cannot spawn more processes (even daemonic ones).
+    # The worker exiting when we get a Ctrl+C is just an added (but necessary) bonus.
+    p = mp.Process(target=_worker_process, args=(fn, dest, args, kwargs, debug), daemon=True)
+    p.start()
+    await join_async(p, timeout, interval)
+    if p.is_alive():
+        p.kill()  # Die!
+        raise TestTimeout(f"Function took too long (exceeded timeout of {timeout}s)")
+    if p.exitcode != 0:
+        raise mp.ProcessError("Error in process") from dest.get()
+    return dest.get()
+
+
+async def run_with_timeout_async(timeout: float, fn, args, kwargs,
+                                 debug=0, interval=0.005, pool: bool | mp.pool.Pool = None):
+    """fn must be pickleable and a regular function **not** coroutine!
+
+    debug:
+      - 0: normal
+      - 1: print pid on start
+      - 2: also wait 15s in process to allow time to 'Attach to Process' in Pycharm
+    Note: don't use debug=2 with low timeouts unless you want to get terminate()d"""
+    if not pool:
+        return await _run_with_timeout_async_process(timeout, fn, args, kwargs, debug, interval)
+    if pool is True:
+        if run_with_timeout_async.default_pool is None:
+            run_with_timeout_async.default_pool = SimpleProcessPool(4)
```

Comment:
TODO: use a more sensible default e.g. `cpus*0.8` or `cpus`

Reply:
Growing pool maybe in a later PR. Right now 4 is a good default (if too many, it takes ages to create the processes on windows - windows processes are SO SLOW!!!).

Reply:
For specifying the size, that will be enough for after/in the growing the pool PR. 
This is so that we don't need to maintain many pools of different size (which would be a ton of RAM - windows processes are insanely heavyweight) and instead we can specify a minimum size and grow pool if required.